wdl-format/src/comments.rs

```diff
@@ -0,0 +1,188 @@
+//! Format comments in a WDL file.
+//!
+//! All comments will be treated as either "preceding" or "inline" comments.
+//! Every comment will "belong" or "be owned" by a specific element in the
+//! syntax tree. If that element is moved from one place to another, the
+//! comment will move with it. Only syntax elements that are either the first
+//! element of a line or the last element of a line can own comments. Elements
+//! may span multiple lines, only the beginning (in the case of preceding
+//! comments) or the end (in the case of inline comments) of the element
+//! are considered.
+//!
+//! A preceding comment is a comment that appears on a line before an element.
+//! There may be any number of preceding comments and they may be separated
+//! by any number of blank lines. Internal blank lines (blanks _between_
+//! comments) will be included in the formatted output, but blank lines before
+//! or after all comments will not be includes. Multiple internal blank lines in
+//! a row will be consolidated to one. Preceding comments should be indented to
+//! the same level as the element which they belong to.
+//!
+//! An inline comment is a comment that appears on the same line as an element,
+//! if and only if that element is the last element of its line. Inline comments
+//! should always appear two spaces after the element they are commenting on.
+
+use wdl_ast::AstToken;
+use wdl_ast::Comment;
+use wdl_ast::SyntaxElement;
+use wdl_ast::SyntaxKind;
+
+use super::Formattable;
+use super::State;
+use super::NEWLINE;
+
+/// Inline comment space constant used for formatting. Inline comments should
+/// start two spaces after the end of the element they are commenting on.
+pub const INLINE_COMMENT_SPACE: &str = "  ";
+
+impl Formattable for Comment {
+    fn format<T: std::fmt::Write>(&self, writer: &mut T, _state: &mut State) -> std::fmt::Result {
+        let comment = self.as_str().trim();
+        write!(writer, "{}{}", comment, NEWLINE)
+    }
+}
+
+/// Format comments that precede a node.
+///
+/// This function assumes we are _not_ in the preamble, and thus any
+/// double-pound-sign comments should be converted to single-pound-sign
+/// comments.
```

Comment:
I think this is appropriate, considering it only checks for `## ` comments (note the space) which should be pretty error-proof. Double-pound comments are _exclusively_ for the preamble, so this feels safer than the reverse operation (upgrading single-pound-comments in the preamble to double-pound).